File: Java/216.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList();
        Helper(1, 0, k, n, res, new ArrayList());
        return res;
    }
    
    private void Helper(int ind, int sum, int k, int n, List<List<Integer>> res, List<Integer> l) {
        if (l.size() == k) {
            if (sum == n) {
                res.add(new ArrayList(l));
            }
            return;
        }
        for (int i = ind; i <= 9; i++) {
            if (i > (n - sum) / (k - l.size()) + 1)
                return;
            else {
                l.add(i);
                Helper(i + 1, sum + i, k, n, res, l);
                l.remove(l.size() - 1);
            }
        }
    }
}
__________________________________________________________________________________________________
sample 32028 kb submission
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(k, n, 1, new ArrayList<>(), res);
        return res;
    }
    
    public void dfs(int k, int target, int startNum, List<Integer> subList, List<List<Integer>> res) {
        if (target == 0 && subList.size() == k) {
            res.add(new ArrayList<>(subList));
            return;
        }
        
        for (int i = startNum; i <= 9; i++) {
            if (i > target) {
                break;
            }
            
            subList.add(i);
            dfs(k, target - i, i + 1, subList, res);
            subList.remove(subList.size() - 1);
        }
    }
}
__________________________________________________________________________________________________

File: Python3/216.py
__________________________________________________________________________________________________
sample 16 ms submission
class Solution:
    def combinationSum3(self, k: int, n: int, d:int = 9) -> List[List[int]]:
        if k * (2 * d - k + 1) <= 2 * n:
            return [list(range(d - k + 1, d + 1))] if k * (2 * d - k + 1) == 2 * n else []
        if k * (k + 1) >= 2 * n:
            return [list(range(1, k + 1))] if k * (k + 1) == 2 * n else []
        return [l + [d] for l in self.combinationSum3(k - 1, n - d, d - 1)] + self.combinationSum3(k, n, d - 1);
__________________________________________________________________________________________________
sample 13000 kb submission
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        
        nums = list(range(1, 10))
        res = []
        
        def dfs(nums, path, k, target, res):
            
            if k < 0 or sum(path) > target:
                return
            
            if k == 0 and sum(path) == target:
                res.append(path)
                return
            
            for i in range(len(nums)):
                
                dfs(nums[i+1:], path + [nums[i]], k-1, target, res)
                
        dfs(nums, [], k, n, res)
    
        return res
__________________________________________________________________________________________________
